### DISCUSSION TITLE
I have used pyinstaller to compile my MicroPython files, now can’t transfer the compiled file onto the Pico

### DISCUSSION THREAD

**Discussion by FlydGitHub** [micropython/micropython#10873] [RP2040 / Pico] [answered]
Hi 

Any suggestions: I have used pyinstaller to compile my MicroPython files, now can’t transfer the compiled file onto the Pico

Thanks

**Comment by FlydGitHub** [comment]
mpremote works just fine on the Raspberry Pi 4B , but not on the pico

**Comment by peterhinch** [comment]
`pyinstaller` is intended to run on a PC and is not relevant to microcontrollers. You need to download and install firmware [from here](https://micropython.org/download/?port=rp2). Then `mpremote` will work and you can use it to install your code.

**Reply by FlydGitHub** [reply]
Thank you for your reply.

I have used  mpremote on my Raspberry Pi 4B -4GB, but unable to use in on the Pico RP2040 w

My systems are 1. Raspberry Pi OS 32 bit current version. 2. MacOS, 3.  Pico RP2040 w

**Reply by peterhinch** [reply]
`mpremote` definitely works with the Pico W. I suspect you might be misunderstanding its use. Firstly you need to be sure your Pico W is running current MicroPython firmware, V1.19.1 or the latest daily build.

`mpremote` can run on any PC: either your Raspberry Pi 4B or on your Mac. Connect the PC to the Pico W with a USB cable, run `mpremote` on the PC, and you should get a REPL.

**Comment by FlydGitHub** [comment]
Thank you Peter,

I very much appreciate your feedback.

mpremote is awesome!   I am running  V1.19.1 or the latest daily build.

I still don't know how to transfer the output from mpremote onto my Pico w?

Thanks

**Comment by FlydGitHub** [comment]
Hi Peter, et all

1. Successfully used mpremote to create a Linux executable file, still unable to transfer the Linux executable,’command line’ / ‘binary' file to either MicroPython or CircuitPython.
They both appear to be only accept text files.
Am I missing something?

2. I have tried to execute it like a shell commands in Python - using Subprocess Run Method, but that  fails with error message:
 
ImportError: cannot import name 'TerminalMunu' from 'simple_term_menu' (/home/pi/.local/lib/python3.9/site-packages/simple_term_menu.py)

3. FYI I just found the Challenger RP2040 witch supports 8MB - Yeah !!
Thanks

FYI Code for Subprocess follows:

simple_term_menu.py

from simple_term_menu import TerminalMunu 
 
options  = ["[a] Apples", "[b] Bannas", "[o] Oranges","[q] Ouitl"]

mainMenu = TerminalMenu(options)

quitting = False
          
while quitting == False:
    optionsIndex = mainMenu.show()
    optionsChoice = options[optionsIndex]
    
    if(optionsChoice == "[q] Quit"):
        quitting = True
    else:
        print("optionsChoice")

**Comment by peterhinch** [comment]
You use `mpremote` to transfer Python source files (text files) to the target as per [the docs](http://docs.micropython.org/en/latest/reference/mpremote.html?highlight=mpremoe).

**Comment by FlydGitHub** [comment]
Thank you Peter,

So how to I transfer non text files to my pico?

**Comment by bixb922** [comment]
mpremote works for both text files and non-text files. Micropython programs are .py text files, such as main.py, and you just copy those files to the microcontroller with mpremote (unless you wish to use the mpy-cross compiler, but that's not necessary for starters).
What non text files are you trying to transfer?
What do you mean by "transfer a Linux executable"? Micropython executes .py (and .mpy) files.

**Comment by FlydGitHub** [comment]
" Linux executable"?
As is a shell command

**Comment by dpgeorge** [comment]
@FlydGitHub you need to clearly explain (1) what you are trying to do; (2) the exact steps you have taken to do that thing; (3) the output you got

**Comment by FlydGitHub** [comment]
Hi Damien,

Thank you for your patience. 

Please forgive me I have a traumatic brain injury, so processing, especially text is challenging.

1. I want to compile my Python code.

2. Call the resulting code inside a second python script, on a pico w. 

Is that any clearer?

Thanks

**Reply by dpgeorge** [reply]
OK, so let's start with the basics.

1. Do you have MicroPython installed on your Pico W?
2. Can you get a REPL prompt on the Pico W using `mpremote` (without any arguments, just run `mpremote`)?
3. Can you run a simple .py script (eg test.py) using mpremote: `mpremote run test.py`?

**Comment by FlydGitHub** [comment]
1.Yes
2. I have mpremote installed. But get:
MicroPython v1.19.1-910-g4937174b4 on 2023-03-01; Raspberry Pi Pico W with RP2040
Type "help()" for more information.
>>> mpremote
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'mpremote' isn't defined
>>>

**Reply by robert-hh** [reply]
mpremote must be started on the PC, not on the Pico board.

**Reply by dpgeorge** [reply]
OK, that looks good, you can run `mpremote` on your PC and get a REPL prompt on the Pico W (I assume that's how you are seeing that banner with the MicroPython version number).

Next, try typing `1/2` at the REPL.  Does it print `0.5` as a result?

Then create a text file called `test.py` that contains `print("hello")`.  Save it.  Then from your PC command line run `mpremote run test.py`.  It should print out "hello".

**Comment by FlydGitHub** [comment]
OK that all works. 

1. Now how do a copy/move test.py into the Pico?
2. I can't seem to find how to call the resulting code inside a second python script, on a pico w.

**Comment by bixb922** [comment]
1. If you have test.py on your PC, you copy it to the Pico with ```mpremote cp test.py :test.py```

2. If test.py has a function called my_func, you call that from another script with
```
from test import my_func
# now you can call my_func:
my_func()
```

**Comment by FlydGitHub** [comment]
Thank you that's great!

1. Now I use pyinstaller to create an ARM executable, I can move the file onto the Pico, but I don't see it.
2. I have just copied the binary, witch is only ~150kbites, will that work, as the one file option give 5Meg!

**Comment by bixb922** [comment]
I don't know if I understand the question well, but ARM executables don't run on Micropython nor do they on the Raspberry Pico. Micropython runs .py and .mpy files

**Comment by FlydGitHub** [comment]
The Pico -  dual-core ARM Cortex-M0+ processor so is that compatible?

I thought the ARM Instruction set was common?

**Comment by FlydGitHub** [comment]
OK so it's not a binary file, my Mac says it's Unix Executable.

**Comment by bixb922** [comment]
If it is a Unix executable, it will need the services the Linux operating system provides. Micropython does not provide all those services, since there is no underlying operating system. There is no Linux on the Pico Micropython.

AFAIK, pyinstaller works with Python but not with Micropython. With Micropython you deploy the .py text files to the Pico and they are compiled there on the fly by Micropython and then executed. There are other more involved ways, for example using mpy-cross or freezing the code in the Micropython image. To get started, running .py text files on the Pico is the easiest way.

**Comment by FlydGitHub** [comment, marked as answer]
Thanks for your awesome support!!
